### PR TITLE
Add new BDVReader

### PR DESCRIPTION
**Pull request by dgault**
BDVReader for the current OME Model, based on https://arxiv.org/pdf/1412.0488.pdf

XML file is read and parsed to locate h5 file and a number of parameters.
H5 file is then read using the jhdf library.

Currently only h5 files supported.

### CODE REVIEW

**Comment by sbesson** [comment]
Tested the latest version of this by importing one of the BDV datasets from `idr0038` into a testing server with this PR included. The import with `bin/omero impoort --transfer=ln_s --skip=checksum --skip=minmax` completed without issue but the thumbnail generation as well as the image opening was broken with the following error in `Blitz.log`:

```
java.lang.RuntimeException: ncsa.hdf.hdf5lib.exceptions.HDF5FileNotFoundException: Path is not a valid HDF5 file. (/home/omero/omero-server-data/ManagedRepository/bdv_302/2018-06/08/10-29-29.803/Lam-Cytok-MV-f0.xml)
	at ome.io.bioformats.BfPixelBuffer.reader(BfPixelBuffer.java:79) ~[romio.jar:na]
...
Caused by: ncsa.hdf.hdf5lib.exceptions.HDF5FileNotFoundException: Path is not a valid HDF5 file. (/home/omero/omero-server-data/ManagedRepository/bdv_302/2018-06/08/10-29-29.803/Lam-Cytok-MV-f0.xml)
	at ch.systemsx.cisd.hdf5.HDF5BaseReader.openFile(HDF5BaseReader.java:195) ~[jhdf5.jar:na]
	at ch.systemsx.cisd.hdf5.HDF5BaseReader.<init>(HDF5BaseReader.java:141) ~[jhdf5.jar:na]
	at ch.systemsx.cisd.hdf5.HDF5BaseReader.<init>(HDF5BaseReader.java:120) ~[jhdf5.jar:na]
	at ch.systemsx.cisd.hdf5.HDF5ReaderConfigurator.reader(HDF5ReaderConfigurator.java:86) ~[jhdf5.jar:na]
	at ch.systemsx.cisd.hdf5.HDF5FactoryProvider$HDF5Factory.openForReading(HDF5FactoryProvider.java:54) ~[jhdf5.jar:na]
	at ch.systemsx.cisd.hdf5.HDF5Factory.openForReading(HDF5Factory.java:64) ~[jhdf5.jar:na]
	at loci.formats.services.JHDFServiceImpl.setFile(JHDFServiceImpl.java:85) ~[formats-gpl.jar:5.9.0-SNAPSHOT]
	at loci.formats.in.BDVReader.initializeJHDFService(BDVReader.java:297) ~[formats-gpl.jar:5.9.0-SNAPSHOT]
	at loci.formats.in.BDVReader.reopenFile(BDVReader.java:284) ~[formats-gpl.jar:5.9.0-SNAPSHOT]
	at loci.formats.ImageReader.reopenFile(ImageReader.java:848) ~[formats-api.jar:5.9.0-SNAPSHOT]
	at loci.formats.ReaderWrapper.reopenFile(ReaderWrapper.java:655) ~[formats-api.jar:5.9.0-SNAPSHOT]
	at loci.formats.ReaderWrapper.reopenFile(ReaderWrapper.java:655) ~[formats-api.jar:5.9.0-SNAPSHOT]
	at loci.formats.Memoizer.setId(Memoizer.java:663) ~[formats-bsd.jar:5.9.0-SNAPSHOT]
	at ome.io.bioformats.BfPixelsWrapper.<init>(BfPixelsWrapper.java:52) ~[romio.jar:na]
	at ome.io.bioformats.BfPixelBuffer.reader(BfPixelBuffer.java:73) ~[romio.jar:na]
	... 72 common frames omitted
```

**Comment by sbesson** [comment]
Initial feedback as I have been testing this reader the  BDV datasets submitted as part of `idr0038`. Filesets were tested via a classical OMERO import worflow.

Filesets detection client-side was tested with `bin/omero import -f --readers readers.txt` where `readers.txt` contained only `loci.formats.in.BDVReader`:

- one of the filesets fails detection with a NPE in `loci.formats.in.BDVReader.initializeJHDFService`
- the average `setId` time on each fileset is quite long i.e. `100000ms` for filesets composed of two files: one XML and one H5 file

Importing one of the detected filesets resulted in two immediate issues:

- 10 images are detected and created but one of them is associated with a plate while the other 9 ones end up in the orphaned folder. What triggers the HCS classification?
- pixel data reading (thumbnails, image opening) fails withe same exeption as the one reported in https://github.com/openmicroscopy/bioformats/pull/3151#issuecomment-395723013 i.e `ncsa.hdf.hdf5lib.exceptions.HDF5FileNotFoundException`

**Comment by dgault** [comment]
The last commit should clear up the exception issues. The plate, well and site metadata is parsed along with a series name, so should be included for each series (with the exception of sub resolutions)

File: components/formats-gpl/src/loci/formats/in/BDVReader.java [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+  /* @see loci.formats.FormatReader#initFile(String) */
+  @Override
+  protected void initFile(String id) throws FormatException, IOException {
+    super.initFile(id);
+    xmlId = id;
+    store = makeFilterMetadata();
+    in = new RandomAccessInputStream(xmlId);
+    in.setEncoding("ASCII");
+    DefaultHandler handler = new BDVXMLHandler();
+    try {
```
Comment (jburel) [inline components/formats-gpl/src/loci/formats/in/BDVReader.java]: This could leak, change recommended
```
try (RandomAccessInputStream s = new RandomAccessInputStream(id)) {
      XMLTools.parseXML(s, handler);
 } catch (IOException e) {
      throw new FormatException("Malformed XML", e);
 } 

```

Comment (joshmoore) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: Perhaps something that could be part of XMLTools?

Comment (jburel) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: From previous discussions I understood that the reader is responsible to close the stream. So I   did not look at changing the underline methods. TIFFParser for example follows the same pattern

Comment (joshmoore) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: I was thinking that another `parseXML` method would take the `id` and then could manage the stream internally.

File: components/formats-gpl/src/loci/formats/in/BDVReader.java [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+                              {255, 255, 0}, {0, 255, 255}, {255, 0, 255},
+                              {255, 255, 255}, {255, 0, 128}, {0, 255, 128},
+                              {0, 128, 256}, {128, 0, 128}, {255, 128, 0},
+                              {64, 128, 0}, {0, 64, 128}, {128, 0, 64}};
+
+  // -- Fields --
+
+  private MetadataStore store;
+  private String bdvxml;
+  String xmlId;
```
Comment (jburel) [inline components/formats-gpl/src/loci/formats/in/BDVReader.java]: Why is that needed? and not using directly id

File: components/formats-gpl/src/loci/formats/in/BDVReader.java [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+  String h5Id;
```
Comment (jburel) [inline components/formats-gpl/src/loci/formats/in/BDVReader.java]: private?

File: components/formats-gpl/src/loci/formats/in/BDVReader.java [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+    try (RandomAccessInputStream s = new RandomAccessInputStream(id)) {
+      XMLTools.parseXML(s, handler);
+    }
+    catch (IOException e) {
+      throw new FormatException("Malformed XML", e);
+    }
+    store.setXMLAnnotationValue(bdvxml, 0);
+    String xml_id = MetadataTools.createLSID("XMLAnnotation", 0); 
+    store.setXMLAnnotationID(xml_id, 0);
+    initializeJHDFService(h5Id);
```
Comment (jburel) [inline components/formats-gpl/src/loci/formats/in/BDVReader.java]: Does it mean that the method ``characters`` has to be invoked before otherwise h5id is not set

File: components/formats-gpl/src/loci/formats/in/BDVReader.java [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+  /* @see loci.formats.IFormatReader#close(boolean) */
+  @Override
+  public void close(boolean fileOnly) throws IOException {
+    super.close(fileOnly);
+    if (!fileOnly) {
+      seriesCount = 0;
+      H5PositionList.clear();
+      H5PathsToImageData.clear();
+      cellObjectNames.clear();
+      if (jhdf != null) {
```
Comment (jburel) [inline components/formats-gpl/src/loci/formats/in/BDVReader.java]: if ``super.close`` throws an exception then ``jhdf`` is never closed

File: components/formats-gpl/src/loci/formats/in/BDVReader.java (hunk 4 above)
Comment (sbesson) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: Reading briefly the code, `DefaultHandler handler = new BDVXMLHandler();` will createsa new BDV Handler and this is where `h5id` gets initialized as per the overriden [DefaultHandler.characters(char\[\], int, int)](http://www.saxproject.org/apidoc/org/xml/sax/helpers/DefaultHandler.html#characters(char[],%20int,%20int)) API.

It is also unclear to me whether there would be conditions under which `h5Id` would not be initialized. In this case, we might want to consider  a NPE check here or in `initializeJHDFService` @dgault ?

Comment (jburel) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: In this case, doc should be added to describe the sequence of calls

**Comment by dgault** [comment]
The recent timepoint changes are to address the exception when importing the remaining samples. The previous samples and the example XML all use timepoints similar to the below:
```
    <Timepoints type="range">
      <first>1</first>
      <last>10</last>
    </Timepoints>
```

The new file that was failing (`HisYFP-SPIM.xml`) has the following:

```
    <Timepoints type="pattern">
      <integerpattern>18</integerpattern>
    </Timepoints>
```

Checking the spec it does mention the option of a pattern or range but I could not find any further details or examples of the pattern. From looking at some of the code I believe the pattern could be as follows:

```
<integerpattern>1-10</integerpattern>
```

Would be the equivalent of a range of timepoints between 1 and 10

```
<integerpattern>1-30:5</integerpattern>
```

Would be the equivalent of a range of timepoints between 1 and 30 but incrementing by 5 between each frame

**Comment by sbesson** [comment]
Retesting specifically on the `HisYFP-SPIM.xml`, a simple call of `showinf` works as expected. However, importing into OMERO fails. This can be more simply reproduced at the Bio-Formats level by calling `setId` with non-flattened resolutions:

```
bf@96c1f66ad887:~$ /bio-formats-build/bioformats/tools/showinf -nopix -noflat /data/public/samples/HisYFP-SPIM.xml 
Checking file format [BDV]
Initializing reader
BDVReader initializing /data/public/samples/HisYFP-SPIM.xml
...
Series #8 :
	Resolutions = 4
		sizeX[0] = 358
		sizeX[1] = 179
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index: 34, Size: 34
	at java.util.ArrayList.rangeCheck(ArrayList.java:657)
	at java.util.ArrayList.get(ArrayList.java:433)
	at loci.formats.FormatReader.currentCore(FormatReader.java:1475)
	at loci.formats.FormatReader.getSizeX(FormatReader.java:662)
	at loci.formats.ImageReader.getSizeX(ImageReader.java:292)
	at loci.formats.ReaderWrapper.getSizeX(ReaderWrapper.java:201)
	at loci.formats.tools.ImageInfo.readCoreMetadata(ImageInfo.java:595)
	at loci.formats.tools.ImageInfo.testRead(ImageInfo.java:1048)
	at loci.formats.tools.ImageInfo.main(ImageInfo.java:1121)
```

In addition to this issue, a couple of feedback testing other sample BDV files:

- the OME-XML validation fails with a few errors

````
bf@96c1f66ad887:~$ /bio-formats-build/bioformats/tools/showinf -nopix -omexml /data/public/samples/drosophila.xml 
Checking file format [BDV]
Initializing reader
BDVReader initializing /data/public/samples/drosophila.xml
Plate :t00000
Well :t00000/s00
Site :t00000/s00/0
Site :t00000/s00/1
Site :t00000/s00/2
Site :t00000/s00/3
Well :t00000/s01
Site :t00000/s01/0
Site :t00000/s01/1
Site :t00000/s01/2
Site :t00000/s01/3
Parsing schema path
http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd
Validating OME-XML
cvc-pattern-valid: Value 'XMLAnnotation:0' is not facet-valid with respect to pattern '(urn:lsid:([\w\-\.]+\.[\w\-\.]+)+:Annotation:\S+)|(Annotation:\S+)' for type 'AnnotationID'.
cvc-attribute.3: The value 'XMLAnnotation:0' of attribute 'ID' on element 'XMLAnnotation' is not valid with respect to its type, 'AnnotationID'.
Error validating document: 2 errors found
Initialization took 16.004s
...
````

- as seen above, the current state of the reader creates a SPW structure. This is most noticable when importing into OMERO as multi-image filesets are partly linked under a Well. I do not know off-hand under which case it is valid to treat a BDV dataset as HCS.
- currently each setup is mapped as a series. In the case of setups defined as follows:

````
<?xml version="1.0" encoding="UTF-8"?>
<SpimData version="0.2">
  <BasePath type="relative">.</BasePath>
  <SequenceDescription>
    <ImageLoader format="bdv.hdf5">
      <hdf5 type="relative">astroP65.h5</hdf5>
    </ImageLoader>
    <ViewSetups>
      <ViewSetup>
        <id>0</id>
        <name>channel 1</name>
        <size>2947 5192 1402</size>
        <voxelSize>
          <unit>pixel</unit>
          <size>1.0 1.0 1.0</size>
        </voxelSize>
        <attributes>
          <channel>1</channel>
        </attributes>
      </ViewSetup>
      <ViewSetup>
        <id>1</id>
        <name>channel 2</name>
        <size>2947 5192 1402</size>
        <voxelSize>
          <unit>pixel</unit>
          <size>1.0 1.0 1.0</size>
        </voxelSize>
        <attributes>
          <channel>2</channel>
        </attributes>
      </ViewSetup>
````

  having each setup being a channel might be a more appropriate representation.

As discussed earlier today, this is typically an example of format reader solely driven by representative samples and as such we should expect multiple versions as new datasets get tested against it. My primary concern is whether we should aim at having our first version available in Bio-Formats 6.0.0. Given the number of discussion points above, my feeling is that we should not block releasing Bio-Formats 6.0.0 on it but target its completion and integration together with other major reader improvements in a Bio-Formats 6.1.0 release shortly after.

**Comment by sbesson** [comment]
Retested the last commits using the BDV samples submitted as part of the `idr0038` and `idr0048` studies:

Screen Shot 2019-04-23 at 09 14 19
Screen Shot 2019-04-23 at 09 12 19

From the representation standpoint, both datasets are correctly detected and imported. The SPW structure is no longer detected as discussed and the image dimensions match the content of the XML with all channel setups detected as `Channel` of the same `Image`.

A few outstanding issues:
- the loading of the data for `idr0048` seem to be off by one-channel and the red channel is identical to the green channels - leading to the yellow spots in the screenshot above. See below for comparison of the BDV dataset (right) and the corresponding OME-TIFF ground truth (left)
Screen Shot 2019-04-23 at 09 12 35
Screen Shot 2019-04-23 at 09 12 44
- the samples from https://imagej.net/BigDataViewer_Example_Datasets both fail to import with a `java.lang.IndexOutOfBoundsException` and a `java.lang.IllegalArgumentException`

Also a few secondary issues
- the initialization logging could probably be downgraded to `DEBUG` level
- the first initialization seemed slow (~10-20s) while most of the structure has been parsed. More logging might be useful to understand where the time is spent for future optimization.

The last two comments are low-priority. Next step is going to start configuring `idr0048` and `idr0038` and add them to the daily CI jobs to start flagging regression. For the progression, equally happy to have the fixes pushed here or to merge this PR which adds most of the logic and have the issues above addressed in separate more contained PRs.

**Comment by sbesson** [comment]
Some initial feedback from retesting this PR using Bio-Formats command-line tools only. All samples now open. I tried to set up automated tests which led to a couple of issues. Some of them are definitely for follow-up PRs but listing here so that we can discuss

- `testSaneUsedFiles` and `testIsThisType` will require additional exclusions to handle the fact a BDV filesets cannot be initialized from the `.h5` file at the moment
- `ID="XMLAnnotation:0"` generates warning during OME-XML validation
- the generated OME-XML fails validation due to the existence of `<Channel>` objects in the `XMLAnnotation`. We might need to have some special handling similar to https://github.com/openmicroscopy/bioformats/blob/v6.1.0-m1/components/formats-api/src/loci/formats/services/OMEXMLServiceImpl.java#L561. Is it time to consider a namespace to skip BDV XMLAnnotation during validation?

I am parallely on getting an OMERO 5 server deployed with this PR included so that we can test the samples via our full application stack.

**Comment by dgault** [comment]
As discussed, added the logic to allow for a h5 extension by searching the same directory for an identically named XML file.

File: components/formats-gpl/src/loci/formats/in/BDVReader.java [hunk 6]
```diff
@@ -0,0 +1,872 @@
...
+            m.indexed = true;
+            int bpp = jhdf.getElementSize(coord.pathToImageData);
+            if (bpp==1) {
+              m.pixelType = FormatTools.UINT8;
+            }
+            else if (bpp==2) {
+              m.pixelType = FormatTools.UINT16;
+            }
+            else if (bpp==4) {
+              m.pixelType = FormatTools.INT32;
```
Comment (mtbc) [inline components/formats-gpl/src/loci/formats/in/BDVReader.java]: Interesting it's not also UINT.

Comment (dgault) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: From the spec that I have the data coming from Big Data Viewer should be UINT16:

```
BigDataViewer always stores images with 16-bit precision currently, while the image you want to export is not necessarily 16-bit. The value range defines the minimum and maximum of the image you want to export. This is mapped to the 16-bit range for export. I.e., the minimum of the value range will be mapped to the minimum of the unsigned 16-bit range (0). The maximum of the value range will be mapped to the maximum of the unsigned 16-bit range (65535).
```

The other options (namely the out of place INT32) are inline with what we have for other H5 files such as CellH5 which has UINT16 and INT32

Comment (sbesson) [reply components/formats-gpl/src/loci/formats/in/BDVReader.java]: Thanks for opening the discussion. The samples we have are all 16-bit as per the specification that David linked from. I think keeping this handling should be largely fine and we can review this if the specification get extended with representative samples. Alternatively, we could be conservative and decide to throw an exception if `bpp` is 4.
